include/onnxruntime/core/session/onnxruntime_c_api_ep.h

```diff
@@ -0,0 +1,649 @@
+
+typedef struct OrtCreateStream {
+  int device_type;
+  void*(ORT_API_CALL* CreateStreamFunc)(const OrtDevice*);
+} OrtCreateStream;
+
+typedef struct OrtMetaDef {
+  char* name;
+  char* domain;
+  int since_version;
+
+  char** inputs;
+  size_t input_len;
+  char** outputs;
+  size_t output_len;
+  char** constant_initializers;
+  size_t initializer_len;
+
+  char* doc_string;
+} OrtMetaDef;
+
+typedef struct OrtIndexedSubGraph {
+  OrtMetaDef* meta_def; // TODO(leca): how to define a nested structure pointer?
+  size_t* node_index;
+  size_t node_index_len;
+} OrtIndexedSubGraph;
+
+typedef struct OrtComputeContext {
+  void*(ORT_API_CALL* AllocateFunc)(void*, size_t, size_t);
+  void(ORT_API_CALL* DestroyFunc)(void*, void*);
+  void* allocator_handle;
+  const char* node_name;
+} OrtComputeContext;
+
+typedef struct OrtNodeComputeInfo {
+  int(ORT_API_CALL* CreateFunctionStateFunc)(OrtComputeContext*, void*, void**);
+  OrtStatusPtr(ORT_API_CALL* ComputeFunc)(void*, void*, const OrtApi*, OrtKernelContext*);
+  void(ORT_API_CALL* DestroyFunctionStateFunc)(void*);
+} OrtNodeComputeInfo;
+
+typedef struct OrtTensorRef {   // TODO(leca): OrtValueInfoRef inside OrtTensorRef?
+  int64_t* shape;
+  size_t shape_len;
+  ONNXTensorElementDataType data_type;
+  const char* data;
+  size_t data_len;
+} OrtTensorRef;
+
+typedef struct OrtValueInfoRef {
+  int64_t* shape;
+  size_t shape_len;
+  ONNXTensorElementDataType data_type;
+} OrtValueInfoRef;
+
+typedef struct OrtExecutionProvider {
+#ifdef __cplusplus
+  OrtExecutionProvider() : GetCapability{nullptr}, Compile{nullptr}, RegisterKernels{nullptr}, CanCopy{nullptr}, CopyTensor{nullptr}, CreatePreferredAllocators{nullptr}, type{nullptr}, create_stream{nullptr}, default_device{nullptr},
+                           extra_param_for_create_state_func{nullptr}, extra_param_for_compute_func{nullptr} {}
+#endif
+  void(ORT_API_CALL* GetCapability)(const OrtExecutionProvider* this_, const OrtGraphViewer* graph, size_t* cnt, OrtIndexedSubGraph***);
+  OrtStatusPtr(ORT_API_CALL* Compile)(OrtExecutionProvider* this_, const OrtGraphViewer** graph, const OrtNode** node, size_t cnt, OrtNodeComputeInfo* node_compute_info);
+  void(ORT_API_CALL* RegisterKernels)(OrtKernelRegistry* kernel_registry);
+  bool(ORT_API_CALL* CanCopy)(const OrtDevice* source, const OrtDevice* target);
+  OrtStatusPtr(ORT_API_CALL* CopyTensor)(const void* src, OrtMemoryInfoDeviceType source_device_type, OrtMemoryType source_mem_type, void* dst, OrtMemoryInfoDeviceType target_device_type, size_t count, void* stream);
+  int(ORT_API_CALL* CreatePreferredAllocators)(OrtExecutionProvider* this_, OrtAllocator*** ort_allocators);
+  void(ORT_API_CALL* ReleaseIndexedSubGraphs)(OrtIndexedSubGraph** indexed_sub_graphs, size_t num_sub_graph);
+  const char* type;
+  OrtCreateStream* create_stream;
+  const OrtDevice* default_device;
+  void* extra_param_for_create_state_func;
+  void* extra_param_for_compute_func;
```

Comment:
I think this may be able to be replaced with a single `void* state;`. See https://github.com/microsoft/onnxruntime/pull/21450/files#r1868553280